x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -330,84 +333,220 @@ public Task createTask(long id, String type, String action, TaskId parentTaskId,
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
         latestState = event.state();
-        if (event.metadataChanged()) {
-            final boolean isResetMode = MlMetadata.getMlMetadata(event.state()).isResetMode();
-            TrainedModelAssignmentMetadata modelAssignmentMetadata = TrainedModelAssignmentMetadata.fromState(event.state());
-            final String currentNode = event.state().nodes().getLocalNodeId();
-            final boolean isNewAllocationSupported = event.state()
-                .getMinTransportVersion()
-                .onOrAfter(TrainedModelAssignmentClusterService.DISTRIBUTED_MODEL_ALLOCATION_TRANSPORT_VERSION);
-
-            if (isResetMode == false && isNewAllocationSupported) {
-                updateNumberOfAllocations(modelAssignmentMetadata);
-            }
+        if (event.metadataChanged() == false) {
+            return;
+        }
+
+        final boolean isResetMode = MlMetadata.getMlMetadata(event.state()).isResetMode();
+        TrainedModelAssignmentMetadata modelAssignmentMetadata = TrainedModelAssignmentMetadata.fromState(event.state());
+        final String currentNode = event.state().nodes().getLocalNodeId();
+        final boolean isNewAllocationSupported = event.state()
+            .getMinTransportVersion()
+            .onOrAfter(TrainedModelAssignmentClusterService.DISTRIBUTED_MODEL_ALLOCATION_TRANSPORT_VERSION);
+        final Set<String> shuttingDownNodes = Collections.unmodifiableSet(event.state().metadata().nodeShutdowns().getAllNodeIds());
 
-            for (TrainedModelAssignment trainedModelAssignment : modelAssignmentMetadata.allAssignments().values()) {
-                RoutingInfo routingInfo = trainedModelAssignment.getNodeRoutingTable().get(currentNode);
-                // Add new models to start loading
-                if (routingInfo != null && isNewAllocationSupported) {
-                    if (routingInfo.getState() == RoutingState.STARTING
-                        && deploymentIdToTask.containsKey(trainedModelAssignment.getDeploymentId())
-                        && deploymentIdToTask.get(trainedModelAssignment.getDeploymentId()).isFailed()) {
-                        // This is a failed assignment and we are restarting it. For this we need to remove the task first.
-                        taskManager.unregister(deploymentIdToTask.get(trainedModelAssignment.getDeploymentId()));
-                        deploymentIdToTask.remove(trainedModelAssignment.getDeploymentId());
+        if (isResetMode == false && isNewAllocationSupported) {
+            updateNumberOfAllocations(modelAssignmentMetadata);
+        }
+
+        for (TrainedModelAssignment trainedModelAssignment : modelAssignmentMetadata.allAssignments().values()) {
+            RoutingInfo routingInfo = trainedModelAssignment.getNodeRoutingTable().get(currentNode);
+            if (routingInfo != null) {
```

Comment:
The functionality for each section largely stayed the same, I refactored it into methods that try to capture the intent of the `if-block` in the method's name.